[BUG, sim] 'chromedriver' executable needs to be in PATH

### To reproduce

In console:

```console
pytest pdr_backend/sim/test/
```

And the bug will emerge:
```text
'chromedriver' executable needs to be in PATH
```

Notes:
- It's happening locally 
- I'm running on a MacBook Pro. 
- Other unit tests don't have issues, eg if I run aimodel/test/

### Traceback

```text
(venv) trentmc@trents-m3: ~/code/pdr-backend $ pytest pdr_backend/sim/test/
============================================================================================================================== test session starts ==============================================================================================================================
platform darwin -- Python 3.12.3, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/trentmc/code/pdr-backend
configfile: pytest.ini
plugins: web3-6.19.0, typeguard-4.2.1, env-1.1.3, dash-2.17.1, asyncio-0.21.1, requests-mock-1.12.1
asyncio: mode=Mode.STRICT
collected 8 items                                                                                                                                                                                                                                                               

pdr_backend/sim/test/test_dash_plots.py .....                                                                                                                                                                                                                             [ 62%]
pdr_backend/sim/test/test_multisim_engine.py .                                                                                                                                                                                                                            [ 75%]
pdr_backend/sim/test/test_sim_engine.py E                                                                                                                                                                                                                                 [ 87%]
pdr_backend/sim/test/test_sim_logger.py .                                                                                                                                                                                                                                 [100%]

==================================================================================================================================== ERRORS =====================================================================================================================================
_______________________________________________________________________________________________________________________ ERROR at setup of test_sim_engine _______________________________________________________________________________________________________________________

self = <selenium.webdriver.chrome.service.Service object at 0x3062836e0>

    def start(self):
        """
        Starts the Service.
    
        :Exceptions:
         - WebDriverException : Raised either when it can't start the service
           or when it can't connect to the service
        """
        try:
            cmd = [self.path]
            cmd.extend(self.command_line_args())
>           self.process = subprocess.Popen(cmd, env=self.env,
                                            close_fds=system() != 'Windows',
                                            stdout=self.log_file,
                                            stderr=self.log_file,
                                            stdin=PIPE,
                                            creationflags=self.creationflags)

venv/lib/python3.12/site-packages/selenium/webdriver/common/service.py:71: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py:1026: in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Popen: returncode: 255 args: ['chromedriver', '--port=63889']>, args = ['chromedriver', '--port=63889'], executable = b'chromedriver', preexec_fn = None, close_fds = True, pass_fds = (), cwd = None
env = environ({'TERM_PROGRAM': 'Apple_Terminal', 'SHELL': '/bin/bash', 'TERM': 'xterm-256color', 'TMPDIR': '/var/folders/35/...S_ALLOW_EXTENSION': 'true', 'PYTEST_CURRENT_TEST': 'pdr_backend/sim/test/test_sim_engine.py::test_sim_engine (setup)'}), startupinfo = None
creationflags = 0, shell = False, p2cread = 10, p2cwrite = 14, c2pread = -1, c2pwrite = 15, errread = -1, errwrite = 15, restore_signals = True, gid = None, gids = None, uid = None, umask = -1, start_new_session = False, process_group = -1

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       restore_signals,
                       gid, gids, uid, umask,
                       start_new_session, process_group):
        """Execute program (POSIX version)"""
    
        if isinstance(args, (str, bytes)):
            args = [args]
        elif isinstance(args, os.PathLike):
            if shell:
                raise TypeError('path-like args is not allowed when '
                                'shell is true')
            args = [args]
        else:
            args = list(args)
    
        if shell:
            # On Android the default shell is at '/system/bin/sh'.
            unix_shell = ('/system/bin/sh' if
                      hasattr(sys, 'getandroidapilevel') else '/bin/sh')
            args = [unix_shell, "-c"] + args
            if executable:
                args[0] = executable
    
        if executable is None:
            executable = args[0]
    
        sys.audit("subprocess.Popen", executable, args, cwd, env)
    
        if (_USE_POSIX_SPAWN
                and os.path.dirname(executable)
                and preexec_fn is None
                and not close_fds
                and not pass_fds
                and cwd is None
                and (p2cread == -1 or p2cread > 2)
                and (c2pwrite == -1 or c2pwrite > 2)
                and (errwrite == -1 or errwrite > 2)
                and not start_new_session
                and process_group == -1
                and gid is None
                and gids is None
                and uid is None
                and umask < 0):
            self._posix_spawn(args, executable, env, restore_signals,
                              p2cread, p2cwrite,
                              c2pread, c2pwrite,
                              errread, errwrite)
            return
    
        orig_executable = executable
    
        # For transferring possible exec failure from child to parent.
        # Data format: "exception name:hex errno:description"
        # Pickle is not used; it is complex and involves memory allocation.
        errpipe_read, errpipe_write = os.pipe()
        # errpipe_write must not be in the standard io 0, 1, or 2 fd range.
        low_fds_to_close = []
        while errpipe_write < 3:
            low_fds_to_close.append(errpipe_write)
            errpipe_write = os.dup(errpipe_write)
        for low_fd in low_fds_to_close:
            os.close(low_fd)
        try:
            try:
                # We must avoid complex work that could involve
                # malloc or free in the child process to avoid
                # potential deadlocks, thus we do all this here.
                # and pass it to fork_exec()
    
                if env is not None:
                    env_list = []
                    for k, v in env.items():
                        k = os.fsencode(k)
                        if b'=' in k:
                            raise ValueError("illegal environment variable name")
                        env_list.append(k + b'=' + os.fsencode(v))
                else:
                    env_list = None  # Use execv instead of execve.
                executable = os.fsencode(executable)
                if os.path.dirname(executable):
                    executable_list = (executable,)
                else:
                    # This matches the behavior of os._execvpe().
                    executable_list = tuple(
                        os.path.join(os.fsencode(dir), executable)
                        for dir in os.get_exec_path(env))
                fds_to_keep = set(pass_fds)
                fds_to_keep.add(errpipe_write)
                self.pid = _fork_exec(
                        args, executable_list,
                        close_fds, tuple(sorted(map(int, fds_to_keep))),
                        cwd, env_list,
                        p2cread, p2cwrite, c2pread, c2pwrite,
                        errread, errwrite,
                        errpipe_read, errpipe_write,
                        restore_signals, start_new_session,
                        process_group, gid, gids, uid, umask,
                        preexec_fn, _USE_VFORK)
                self._child_created = True
            finally:
                # be sure the FD is closed no matter what
                os.close(errpipe_write)
    
            self._close_pipe_fds(p2cread, p2cwrite,
                                 c2pread, c2pwrite,
                                 errread, errwrite)
    
            # Wait for exec to fail or succeed; possibly raising an
            # exception (limited in size)
            errpipe_data = bytearray()
            while True:
                part = os.read(errpipe_read, 50000)
                errpipe_data += part
                if not part or len(errpipe_data) > 50000:
                    break
        finally:
            # be sure the FD is closed no matter what
            os.close(errpipe_read)
    
        if errpipe_data:
            try:
                pid, sts = os.waitpid(self.pid, 0)
                if pid == self.pid:
                    self._handle_exitstatus(sts)
                else:
                    self.returncode = sys.maxsize
            except ChildProcessError:
                pass
    
            try:
                exception_name, hex_errno, err_msg = (
                        errpipe_data.split(b':', 2))
                # The encoding here should match the encoding
                # written in by the subprocess implementations
                # like _posixsubprocess
                err_msg = err_msg.decode()
            except ValueError:
                exception_name = b'SubprocessError'
                hex_errno = b'0'
                err_msg = 'Bad exception data from child: {!r}'.format(
                              bytes(errpipe_data))
            child_exception_type = getattr(
                    builtins, exception_name.decode('ascii'),
                    SubprocessError)
            if issubclass(child_exception_type, OSError) and hex_errno:
                errno_num = int(hex_errno, 16)
                if err_msg == "noexec:chdir":
                    err_msg = ""
                    # The error must be from chdir(cwd).
                    err_filename = cwd
                elif err_msg == "noexec":
                    err_msg = ""
                    err_filename = None
                else:
                    err_filename = orig_executable
                if errno_num != 0:
                    err_msg = os.strerror(errno_num)
                if err_filename is not None:
>                   raise child_exception_type(errno_num, err_msg, err_filename)
E                   FileNotFoundError: [Errno 2] No such file or directory: 'chromedriver'

/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py:1955: FileNotFoundError

During handling of the above exception, another exception occurred:

request = <SubRequest 'dash_duo' for <Function test_sim_engine>>, dash_thread_server = <dash.testing.application_runners.ThreadedRunner object at 0x312a53fe0>
tmpdir = local('/private/var/folders/35/dnfp3w4d4zdc5hr8_sllglqh0000gp/T/pytest-of-trentmc/pytest-44/test_sim_engine0')

    @pytest.fixture
    def dash_duo(request, dash_thread_server, tmpdir) -> DashComposite:
>       with DashComposite(
            dash_thread_server,
            browser=request.config.getoption("webdriver"),
            remote=request.config.getoption("remote"),
            remote_url=request.config.getoption("remote_url"),
            headless=request.config.getoption("headless"),
            options=request.config.hook.pytest_setup_options(),
            download_path=tmpdir.mkdir("download").strpath,
            percy_assets_root=request.config.getoption("percy_assets"),
            percy_finalize=request.config.getoption("nopercyfinalize"),
            pause=request.config.getoption("pause"),
        ) as dc:

venv/lib/python3.12/site-packages/dash/testing/plugin.py:179: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.12/site-packages/dash/testing/composite.py:6: in __init__
    super().__init__(**kwargs)
venv/lib/python3.12/site-packages/dash/testing/browser.py:68: in __init__
    self._driver = until(self.get_webdriver, timeout=1)
venv/lib/python3.12/site-packages/dash/testing/wait.py:17: in until
    res = wait_cond()
venv/lib/python3.12/site-packages/dash/testing/browser.py:461: in get_webdriver
    return getattr(self, f"_get_{self._browser}")()
venv/lib/python3.12/site-packages/dash/testing/browser.py:502: in _get_chrome
    else webdriver.Chrome(options=options)
venv/lib/python3.12/site-packages/selenium/webdriver/chrome/webdriver.py:70: in __init__
    super(WebDriver, self).__init__(DesiredCapabilities.CHROME['browserName'], "goog",
venv/lib/python3.12/site-packages/selenium/webdriver/chromium/webdriver.py:89: in __init__
    self.service.start()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <selenium.webdriver.chrome.service.Service object at 0x3062836e0>

    def start(self):
        """
        Starts the Service.
    
        :Exceptions:
         - WebDriverException : Raised either when it can't start the service
           or when it can't connect to the service
        """
        try:
            cmd = [self.path]
            cmd.extend(self.command_line_args())
            self.process = subprocess.Popen(cmd, env=self.env,
                                            close_fds=system() != 'Windows',
                                            stdout=self.log_file,
                                            stderr=self.log_file,
                                            stdin=PIPE,
                                            creationflags=self.creationflags)
        except TypeError:
            raise
        except OSError as err:
            if err.errno == errno.ENOENT:
>               raise WebDriverException(
                    "'%s' executable needs to be in PATH. %s" % (
                        os.path.basename(self.path), self.start_error_message)
E                       selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable needs to be in PATH. Please see https://chromedriver.chromium.org/home

venv/lib/python3.12/site-packages/selenium/webdriver/common/service.py:81: WebDriverException
============================================================================================================================ short test summary info ============================================================================================================================
ERROR pdr_backend/sim/test/test_sim_engine.py::test_sim_engine - selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable needs to be in PATH. Please see https://chromedriver.chromium.org/home
```

--- a/pdr_backend/sim/test/conftest.py
+++ b/pdr_backend/sim/test/conftest.py
@@ -25,7 +25,11 @@ def check_chromedriver():
             It seems that you don't have chromedriver installed on your system.
             Please install it by following the instructions at https://sites.google.com/a/chromium.org/chromedriver/.
             For ubuntu, you can install it using the following command: sudo apt-get install chromium-chromedriver.
-            For Mac, you can install it using the following command: brew cask install chromedriver.
+            For Mac:
+                - you can install it using the following command: brew install chromedriver.
+                - go to chromedriver location and move it to /usr/local/bin/ using this command: sudo mv chromedriver /usr/local/bin/
+                - change the file access mode to add execution permision using this command: sudo chmod +x /usr/local/bin/chromedriver
+                - if a popup appears when running the tests, click on open and test should pass from now on
             For Windows, you can download the chromedriver from the link above and add it to your PATH.
             If you have chromedriver installed, make sure it is in your PATH.
         """

--- a/pdr_backend/sim/test/conftest.py
+++ b/pdr_backend/sim/test/conftest.py
@@ -1,4 +1,11 @@
+import logging
+import pytest
+
+from selenium import webdriver
 from selenium.webdriver.chrome.options import Options
+from selenium.common.exceptions import WebDriverException
+
+logger = logging.getLogger("dependencies")
 
 
 def pytest_setup_options():
@@ -6,3 +13,21 @@ def pytest_setup_options():
     options.add_argument("--headless")
     options.add_argument("--disable-gpu")
     return options
+
+
+@pytest.fixture
+def check_chromedriver():
+    try:
+        driver = webdriver.Chrome()
+        driver.quit()
+    except WebDriverException:
+        message = """
+            It seems that you don't have chromedriver installed on your system.
+            Please install it by following the instructions at https://sites.google.com/a/chromium.org/chromedriver/.
+            For ubuntu, you can install it using the following command: sudo apt-get install chromium-chromedriver.
+            For Mac, you can install it using the following command: brew cask install chromedriver.
+            For Windows, you can download the chromedriver from the link above and add it to your PATH.
+            If you have chromedriver installed, make sure it is in your PATH.
+        """
+
+        logger.error(message)

--- a/pdr_backend/sim/test/test_sim_engine.py
+++ b/pdr_backend/sim/test/test_sim_engine.py
@@ -17,7 +17,8 @@ from pdr_backend.sim.sim_engine import SimEngine
 
 
 @enforce_types
-def test_sim_engine(tmpdir, dash_duo):
+# pylint: disable=unused-argument
+def test_sim_engine(tmpdir, check_chromedriver, dash_duo):
     s = fast_test_yaml_str(tmpdir)
     ppss = PPSS(yaml_str=s, network="development")
 

